blueprints/varnish-support.md

```diff
@@ -0,0 +1,117 @@
+
+Note: the changes should not affect existing components but rather build on them.
+
+### Traffic Portal Impact
+
+n/a
+
+### Traffic Ops Impact
+
+n/a
+
+#### REST API Impact
+
+n/a
+
+#### Client Impact
+
+n/a
+
+#### Data Model / Database Impact
+
+- A new profile type will be added `VARNISH_PROFILE` but it won't affect the model or other structures in `lib/go-tc`
+- `DeliveryService` structs contain fields related to ATS like `remapText`. It will be parsed and translated to Varnish configuration.
+
+### ORT Impact
+
+- `go-varnish` package will be developed to handle generating configuration files for Varnish, Hitch, `varnishncsa` and `logrotate`. For detailed description of mapping configuration files from ATS to Varnish refer to [Varnish Support](https://github.com/apache/trafficcontrol/wiki/Varnish-Support) wiki.
+- New subcommands will be added to `t3c` including `t3c-varnish-generate` that works similar to `t3c-generate` but instead will depend on `go-varnish` package to generate different configuration files and `t3c-varnish-apply` that handles applying these configuration files, restarting and reloading different services running.
+- `go-atscfg` will be refactored to export some of its functionality to be reusable from `go-varnish`. So, instead of rewriting the logic of which IPs are allowed for specific HTTP requests, it could be separated and exported in a function that both packages utilize.
+
+### Traffic Monitor Impact
+
+New statistics parser will be added to Traffic Monitor to handle data coming from Varnish cache statistics endpoint. There is no `VMOD` that exposes Varnish statistics so a service that keeps polling `varnishstat` will be developed.
+
+### Traffic Router Impact
+
+n/a
+
+### Traffic Stats Impact
+
+n/a
+
+### Traffic Vault Impact
+
+n/a
+
+### Documentation Impact
+
+New documentation will be needed for how to setup Varnish with TC and what is the differences between Varnish and ATS.
+
+### Testing Impact
+
+In addition to unit tests and integration tests, [`varnishtest`](https://varnish-cache.org/docs/trunk/reference/varnishtest.html) could be used to test Varnish cache is operating as expected.
+
+### Performance Impact
+
+For current components there should be no performance impact. However, between Traffic Server and Varnish it isn't clear yet what the difference in performance will be.
+
+### Security Impact
+
+n/a
+
+### Upgrade Impact
+
+n/a
+
+### Operations Impact
+
+n/a
+
+### Developer Impact
+
+n/a
+
+## Alternatives
+
+n/a
+
+## Dependencies
+
+- Varnish and its utilities (`varnishtest`, `varnishstat`, `varnishncsa`, ...).
+- Hitch to manage SSL part of the server.
+- `logrotate` to manage logs generated from `varnishncsa`.
```

Comment:
This seems out-of-scope for ATC - some things use logrotate and some don't, but it's not part of our project's responsibilities to dictate that. It's an operations decision how that's handled.

Also, not sure what "manage SSL part of the server" means. SSL keys are retrieved from Traffic Vault, supporting dynamically generating them on the cache-side seems like a headache we don't need.

Reply:
For `logrotate` part, the idea was that ATS already handles log rotating https://docs.trafficserver.apache.org/admin-guide/files/logging.yaml.en.html#logs
So, the idea was to replicate exactly how ATS works in a system using Varnish. However, if it is not desirable to have the cache server handles the log rotation we could ignore that and assume the system administrator will handle that.

For Hitch part, maybe the wording wasn't clear (will change that) but basically it handles secure connections coming to the cache server as Varnish alone does not handle that. So, terminating certificates for example and anything related to `https` requests will be managed by Hitch not Varnish as again it does not support any of that.

Reply:
Varnish really doesn't support HTTPS. I didn't believe you, but then I read the docs. That's.... odd. But regardless, since HTTPS **is** widely used and quite important to ATC instances, it makes sense to accept responsibility for running the TLS proxy - it's basically just a part of the caching service.

I stand by the `logrotate` thing, though.